master/buildbot/test/util/misc.py

```diff
@@ -73,7 +73,13 @@ class TestReactorMixin:
     Mix this in to get TestReactor as self.reactor which is correctly cleaned up
     at the end
     """
-    def setUpTestReactor(self):
+    def setUpTestReactor(self, use_asyncio=False):
+        if use_asyncio:
+            import asyncio
+            import twisted.internet
+            from twisted.internet import asyncioreactor
+            twisted.internet.reactor = asyncioreactor.AsyncioSelectorReactor(asyncio.get_event_loop())
```

Comment:
Shouldn't the code in question use `self.reactor` and thus we should patch that one appropriately? Would it be possible to create fake asyncio reactor?

Reply:
I tried a few options to do that, but the asyncio event loop is much more complex that the twisted one. There are tons of stuff to implement, especially the task manager which schedules the coroutine looks very complex to do.

this post talks a bit about that, and suggests that nobody has yet written an event loop using twisted semantics. 
https://meejah.ca/blog/python3-twisted-and-asyncio

We only have the opposite, i.e twisted loop written in asyncio

I also tried to convert graphql to twisted, but it is also too much hack.

The main drawback of twisted here is that the coroutine support is not equivalent.
you cannot await another coroutine you have to wrap it with ensure-defered.

Reply:
@meejah Do you have some advice on this point? In Buildbot, our test suite is quite fast because we fake the reactor on most unit tests. 

As we start to integrate with asyncio libraries, we must install the asyncioreactor, which takes 60ms per test to initialise( nearly 60x decrease in perf) 

graphql-core merely only uses coroutines and asyncio.gather. Do you think it is worth to try and implement a fake asyncio loop with only support for those?

Reply:
So this wasn't very obvious at first glance, but the asyncio task scheduler only requires loop.call_soon to be implemented. That was quite easy to translate into twisted callLater(0, ...)

Reply:
Okay so graphql-core is written against asyncio .. so you need "something" that lets you inter-op. I think if you want Twisted as well, the best case is still `asyncioreactor`. I don't know what the state-of-the-art for asyncio testing is .. I guess you mean implementing something like `MemoryReactor` / `iosim.IOPump` (but only with the APIs required for graphql to function)...?

Or are you "faking out the reactor" some other way for the Twisted tests?

Reply:
actually graphql-core does not handle the IO part of graphql. It just handle the parsing and resolving. For the latter part there is a little bit of coroutines and asyncio.gather.

The problem of asyncioreactor is that I can see that there seems to be some limitation on windows, which requires to use the non default loop. Windows users have enough problems, I would like to avoid more.

I eventually managed to make the thing work with this solution:
https://github.com/buildbot/buildbot/pull/5980/commits/60c6fd7de28d58986d032eb0e7aeb4cb24140226

This involves implementing a custom asyncio event loop which only implements the call_soon api, which is actually the only thing needed for coroutine scheduling to work.

Reply:
Ahhh, right .. and `asyncio` ties all futures/tasks/etc to an event-loop unlike `Deferred` .. this solution looks decent to me (although I can't claim to have looked in detail).